File: ADFRUITS.py
v1 = [False] * 107
v2 = [False] * 107
arr = [[0 for j in range(107)] for i in range(107)]

def pre():
	for i in range(n):
		v1[i] = False
	for j in range(m):
		v2[j] = False
	for i in range(n+1):
		arr[i][0] = 0
	for i in range(m+1):
		arr[0][i] = 0
	for i in range(1, n+1):
		for j in range(1, m+1):
			if s1[i-1] == s2[j-1]:
				arr[i][j] = arr[i-1][j-1] + 1
			else:
				arr[i][j] = max(arr[i-1][j], arr[i][j-1])
	x, y = n, m
	while x > 0 and y > 0:
		if s1[x-1] == s2[y-1]:
			v1[x-1] = True
			v2[y-1] = True
			x -= 1
			y -= 1
		else:
			if arr[x-1][y] >= arr[x][y-1]:
				x -= 1
			else:
				y -= 1

while True:
	try:
		s1, s2 = input().split()
	except:
		break
	n, m = len(s1), len(s2)
	pre()
	res = list()
	i, j = 0, 0
	while i < n and j < m:
		k, l = i, j
		while k < n and not v1[k]:
			res.append(s1[k])
			k += 1
		while l < m and not v2[l]:
			res.append(s2[l])
			l += 1
		while k < n and l < m and v1[k] and v2[l] and s1[k] == s2[l]:
			res.append(s1[k])
			k += 1
			l += 1
		i, j = k, l
	while i < n:
		res.append(s1[i])
		i += 1
	while j < m:
		res.append(s2[j])
		j += 1
	print(*res, sep = "")

File: ADFRUITS.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

string s1, s2;
int n, m;
bool v1[107], v2[107];
int arr[107][107];

void pre() {
	fill(v1, v1 + n, false);
	fill(v2, v2 + m, false);
	for (int i = 0; i <= n; i++)
		arr[i][0] = 0;
	for (int i = 0; i <= m; i++)
		arr[0][i] = 0;
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= m; j++) {
			if (s1[i-1] == s2[j-1])
				arr[i][j] = arr[i-1][j-1] + 1;
			else
				arr[i][j] = max(arr[i-1][j], arr[i][j-1]);
		}
	}
	int x = n, y = m;
	while (x > 0 && y > 0) {
		if (s1[x-1] == s2[y-1]) {
			v1[x-1] = true;
			v2[y-1] = true;
			x--, y--;
		}
		else
			(arr[x-1][y] >= arr[x][y-1]) ? x-- : y--;
	}
}

int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

#ifndef ONLINE_JUDGE
	freopen("input.txt", "r", stdin);
//	freopen("output.txt", "w", stdout);
#endif

	while (cin >> s1) {
		cin >> s2;
		string res = "";
		n = s1.size(), m = s2.size();
		pre();
		int i = 0, j = 0;
		while (i < n && j < m) {
			int k = i, l = j;
			for ( ; k < n && !v1[k]; k++)
				res += s1[k];
			for ( ; l < m && !v2[l]; l++)
				res += s2[l];
			while (k < n && l < m && v1[k] && v2[l] && s1[k] == s2[l]) {
				res += s1[k];
				k++, l++;
			}
			i = k;
			j = l;
		}
		while (i < n)
			res += s1[i++];
		while (j < m)
			res += s2[j++];
		cout << res << endl;
	}
}
